src/main/java/com/arpnetworking/metrics/mad/sources/TransformingSource.java

```diff
@@ -0,0 +1,429 @@
+                                        getModifiedDimensions(record.getDimensions(), Collections.emptyMap(), ImmutableList.of()));
+                            } else {
+                                final String newMetricName = replacedString.substring(0, tagsStart);
+                                final Map<String, String> parsedTags = TAG_SPLITTER.split(replacedString.substring(tagsStart + 1));
+                                merge(
+                                        metric.getValue(),
+                                        newMetricName,
+                                        mergedMetrics,
+                                        getModifiedDimensions(record.getDimensions(), parsedTags, consumedDimensions));
+                            }
+                        }
+                        //Having "found" set here means that mapping a metric to an empty list suppresses that metric
+                        found = true;
+                    }
+                }
+                if (!found) {
+                    merge(
+                            metric.getValue(),
+                            metricName,
+                            mergedMetrics,
+                            getModifiedDimensions(record.getDimensions(), Collections.emptyMap(), ImmutableList.of()));
+                }
+            }
+
+            // Raise the merged record event with this source's observers
+            // NOTE: Do not leak instances of MergingMetric since it is mutable
+            for (Map.Entry<ImmutableMap<String, String>, Map<String, MergingMetric>> entry : mergedMetrics.entrySet()) {
+                _source.notify(
+                        ThreadLocalBuilder.build(
+                                DefaultRecord.Builder.class,
+                                b1 -> b1.setMetrics(
+                                        entry.getValue().entrySet().stream().collect(
+                                                ImmutableMap.toImmutableMap(
+                                                        Map.Entry::getKey,
+                                                        e -> ThreadLocalBuilder.clone(
+                                                                e.getValue(),
+                                                                DefaultMetric.Builder.class))))
+                                        .setId(record.getId())
+                                        .setTime(record.getTime())
+                                        .setAnnotations(record.getAnnotations())
+                                        .setDimensions(entry.getKey())));
+            }
+        }
+
+        private ImmutableMap<String, String> getModifiedDimensions(
+                final ImmutableMap<String, String> inputDimensions,
+                final Map<String, String> add,
+                final ImmutableList<String> remove) {
+            final Map<String, String> finalTags = Maps.newHashMap();
```

Comment:
You should probably initialize the expected size.

Reply:
or else use a constructor that accepts `inputDimensions`

Reply:
Fixed